test/test_goal_queueing.cpp

```diff
@@ -0,0 +1,305 @@
+	move_base_msgs::MoveBaseGoalConstPtr received_goal;
+
+	// Signaling variables
+	bool finish_executing;
+	bool resume_executing;
+	bool execute_done;
+	std::mutex execute_lock;
+	std::mutex execute_done_lock;
+	std::condition_variable execute_cv;
+	std::condition_variable execute_done_cv;
+
+	std::unique_ptr<actionlib::QueuedActionServer<move_base_msgs::MoveBaseAction>> qserv;
+	std::unique_ptr<actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction>> cli;
+};
+
+// IMPORTANT: Time delays and ros::spinOnce() sequence should stay the same, otherwise, strange things happen
+TEST_F(GoalQueueSuite, establishDuplex) {
+	move_base_msgs::MoveBaseGoal goal; 
+	goal.target_pose.pose.position.x = 3.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear 
+	finishExecuting();
+	ASSERT_TRUE(got_goal);
+	ASSERT_FALSE(goal_preempted);
+	ASSERT_FALSE(next_goal_available);
+	ASSERT_EQ(3.0, received_goal->target_pose.pose.position.x);
+}
+
+TEST_F(GoalQueueSuite, addGoalWhileExecuting) { 
+	move_base_msgs::MoveBaseGoal goal; 
+	goal.target_pose.pose.position.x = 3.0;
+	
+	// First goal
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear 
+	ASSERT_TRUE(qserv->isActive());
+	ASSERT_TRUE(got_goal);
+	ASSERT_EQ(3.0, received_goal->target_pose.pose.position.x);
+
+	// Cancelling the first goal - PITFALL
+	resumeExecuting();
+	cli->cancelGoal();
+	ros::Duration(1.0).sleep(); 
+	ros::spinOnce(); 
+	ASSERT_TRUE(goal_preempted);
+	// Finish the preempted goal
+	finishExecuting();
+	
+	// "Second" goal
+	goal.target_pose.pose.position.x = 7.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear
+	ASSERT_TRUE(qserv->isActive());
+	ASSERT_TRUE(got_goal);
+	ASSERT_EQ(7.0, received_goal->target_pose.pose.position.x); 
+	
+	// Third goal
+	goal.target_pose.pose.position.x = 13.0;
+	cli->sendGoal(goal);
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear
+	resumeExecuting();
+	// Make sure that the "second" goal is still executing, but a new goal is seen
+	ASSERT_TRUE(qserv->isActive()); 
+	ASSERT_TRUE(got_goal);
+	ASSERT_EQ(7.0, received_goal->target_pose.pose.position.x);
+	ASSERT_FALSE(goal_preempted);
+	ASSERT_TRUE(next_goal_available); // Fails when changing other code 
+	// Finish the "second" goal, then the 3nd goal should start executing
+	finishExecuting();
+	ros::spinOnce(); 
+	ros::Duration(0.5).sleep(); // TODO: Make this disappear
+	ASSERT_TRUE(qserv->isActive()); 
+	ASSERT_TRUE(got_goal);
+	ASSERT_FALSE(goal_preempted);
+	ASSERT_FALSE(next_goal_available);
+	ASSERT_EQ(13.0, received_goal->target_pose.pose.position.x);
+	finishExecuting(); // Finish the 3nd goal
+/*
+	- if another goal is received add it to the queue (DONE) 
+	- if the queue full, set the current goal as preempted - explicitly called! - so cancelling the current goal and start executing the next one
```

Comment:
We decided to change this behavior. If the queue is full, the last goal in the queue will be canceled and replaced with the new goal.